### PR TITLE
Relaxing validation of aligned vectorization (part 1)

### LINKED ISSUE #1506: False vectorization validation
When inner-most domains are indexed as contig domains, vectorization validation should be done with the extent of the merged domain. For example, this fusion should allow vectorization by 4 but not:

```
TEST_F(NVFuserTest, FusionVectorizeContigIndex_CUDA) {
  std::vector<int64_t> shape{14, 14};

  Fusion fusion;
  FusionGuard fg(&fusion);

  auto tv0 = makeContigTensor(2);
  fusion.addInput(tv0);
  auto tv1 = set(tv0);
  auto tv2 = set(tv1);
  fusion.addOutput(tv2);

  tv2->merge(0);

  // Vectorize by 4 should be allowed
  tv2->split(0, 4);
  // Vectorize by 2 works
  //tv2->split(0, 2);

  tv2->axis(0)->parallelize(ParallelType::TIDx);
  tv0->computeAt(tv2, 1);

  tv1->axis(1)->parallelize(ParallelType::Vectorize);
  tv2->axis(1)->parallelize(ParallelType::Vectorize);

  fusion.printMath();
  fusion.printKernel();

  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
  at::manual_seed(0);
  auto t0 = at::randn(shape, options);

  FusionExecutor fe;
  fe.compileFusion(&fusion, {t0});
  auto cg_outputs = fe.runFusion({t0});

  TORCH_CHECK(t0.equal(cg_outputs[0]));
}
```
```
unknown file: Failure
C++ exception with description "canVectorize(inputs[it.first], it.second) INTERNAL ASSERT FAILED at "../torch/csrc/jit/codegen/cuda/executor_utils.cpp":618, please report a bug to PyTorch. Error vectorizing, T0_g[ iS13{( ceilDiv(( T0.size[0] * T0.size[1] ), 4) )}, iS14{4} ] as input provided does not allowed vectorization by word size, 4
Exception raised from validateVectorizedTensors at ../torch/csrc/jit/codegen/cuda/executor_utils.cpp:618 (most recent call first):
```

### PR DESCRIPTION
**Pull request by naoyam**
Fixes #1506

The main change is how to validate divisibility of domain extents by vectorization width. Currently, it checks only the inner-most root domain whose extent is not 1. This can be too restrictive as seen in #1506.

In this PR, the validation of divisibility is done against merged contig domains if applicable. Finding such merged domains is done at the time of the vectorization validation of the lowering phase, and the collected information is saved as part of the kernel summary.

See `FusionVectorizeContigIndex` for a concrete example.

I also tried to clean up how info on vectorization is held. We can probably just have all the vectorization information in the kernel summary instead of using the executor cache system, but I also tried to not change so much.

We also need to update the schedulers to allow wider vectorizations, which I plan to work on as a follow-up PR.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/contiguity.h [hunk 1]
```diff
@@ -0,0 +1,83 @@
+#pragma once
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/contiguity.h:1]: Note: this looks exactly copied over from index_compute.cpp

File: torch/csrc/jit/codegen/cuda/contiguity.cpp [hunk 2]
```diff
@@ -0,0 +1,165 @@
+#include <torch/csrc/jit/codegen/cuda/ir_utils.h>
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/contiguity.cpp:1]: Note: I don't see any changes here, just copied directly from index_compute.cpp

File: torch/csrc/jit/codegen/cuda/executor_utils.cpp [hunk 3]
```diff
@@ -368,284 +375,272 @@ void validateKernelOutputs(
...
+    TensorView* consumer_tv,
+    Fusion* fusion) {
+  const bool is_producer_contig = std::all_of(
+      producer_tv->domain()->contiguity().begin(),
+      producer_tv->domain()->contiguity().end(),
+      [](bool contig) { return contig; });
+  const bool is_consumer_contig = std::all_of(
+      consumer_tv->domain()->contiguity().begin(),
+      consumer_tv->domain()->contiguity().end(),
+      [](bool contig) { return contig; });
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/executor_utils.cpp]: Nit: Naming `is_contig` -> `is_fully_contig`.

File: torch/csrc/jit/codegen/cuda/executor_utils.cpp [hunk 4]
```diff
@@ -368,284 +375,272 @@ void validateKernelOutputs(
...
+
+  // When the producer is a fusion input, validate only the producer
+  // and assume the consumer is contiguous. Similarly, when the
+  // consumer is a fusion output, validate the consumer and assume the
+  // producer is contiguous.
+
+  if (producer_tv->isFusionInput()) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/executor_utils.cpp:393]: So if producer is fusion input and consumer is fusion output the output contiguity will be ignored?

File: torch/csrc/jit/codegen/cuda/lower2device.cpp [hunk 5]
```diff
@@ -274,6 +267,10 @@ void GpuLower::lower(Fusion* fusion, DataType index_type) {
   // all IterDomains
   haloInfo().build(fusion_);
 
+  // Want to run this after parallel map and halo info map are
+  // created
+  validateVectorize(fusion_);
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower2device.cpp]: Since there's no build semantic can you mention in the note that it's filling `vectorized_set_info`? Otherwise it might be nice to follow the `.build(fusion_)` pattern.

File: torch/csrc/jit/codegen/cuda/lower_validation.cpp [hunk 6]
```diff
@@ -260,6 +261,97 @@ class VectorizeValidator : public OptInDispatch {
...
+        consumer_contig_finder.rootToContigID().find(consumer_root_id);
+    TORCH_INTERNAL_ASSERT(
+        consumer_contig_it != consumer_contig_finder.rootToContigID().end(),
+        "Contiguity information not found for root domain: ",
+        consumer_root_id->toString());
+    auto consumer_contig_id = consumer_contig_it->second;
+    auto consumer_within_contig_it =
+        consumer_contig_finder.withinContigIDs().find(consumer_contig_id);
+    const auto& consumer_within_contig_root = consumer_within_contig_it !=
+            consumer_contig_finder.withinContigIDs().end()
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_validation.cpp]: Is this not guaranteed by...?
```.cpp
auto consumer_contig_it =
        consumer_contig_finder.rootToContigID().find(consumer_root_id);
```

File: torch/csrc/jit/codegen/cuda/lower_validation.cpp [hunk 7]
```diff
@@ -260,6 +261,97 @@ class VectorizeValidator : public OptInDispatch {
...
+      TensorView* consumer_tv,
+      VectorizedSetInfo& info) {
+    auto producer_tv =
+        consumer_tv->definition()->inputs().at(0)->as<TensorView>();
+
+    const auto c2p = PairwiseRootDomainMap(producer_tv, consumer_tv)
+                         .mapConsumerToProducer(
+                             consumer_tv->domain(), producer_tv->domain());
+
+    auto consumer_root_id = info.vectorized_root_id;
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_validation.cpp]: Could you just note:
```
    // info.vectorized_root_id is validated at this point to be the last
    // concrete root domain in consumer.
```

**Review by csarofeen — COMMENTED** [review]
Do we have failing cases in our tests for vectorization validation? Just want to make sure the paths still fail correctly as well as succeeding in the new case.

File: torch/csrc/jit/codegen/cuda/contiguity.h (hunk 1 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/contiguity.h:1]: Yes, it's now used from the vectorization validation, so I needed to pull it out.

File: torch/csrc/jit/codegen/cuda/contiguity.cpp (hunk 2 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/contiguity.cpp:1]: yes

**Comment by naoyam** [comment]
> Do we have failing cases in our tests for vectorization validation? Just want to make sure the paths still fail correctly as well as succeeding in the new case.

`FusionVectorization2` and `FusionVectorization3` do have `ASSERT_ANY_FAILURE` to make sure assertions get triggered. I'll also extend the new test to make sure lowering should fail without contig merge.

File: torch/csrc/jit/codegen/cuda/executor_utils.cpp (hunk 4 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/executor_utils.cpp:393]: I didn't think about directly doing vectorized copy from an input to an output. Will address such patterns.

File: torch/csrc/jit/codegen/cuda/lower_validation.cpp (hunk 6 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_validation.cpp]: Not actually. `within_contig_ids` is only defined for contig IDs, but `root_to_contig_id_` is defined for all root IDs, which just maps to itself for a non-contig root. I realized it's rather inconsistent.

File: torch/csrc/jit/codegen/cuda/executor_utils.cpp (hunk 4 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/executor_utils.cpp:393]: Added a new test to copy input to output with vectorization.

File: torch/csrc/jit/codegen/cuda/contiguity.h (hunk 1 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/contiguity.h:1]: Yeah, that was expected to be clear, I just needed to comb through it to make sure there wasn't any changes, same as below.

File: test/cpp/jit/test_gpu.cpp [hunk 8]
```diff
@@ -21648,6 +21648,110 @@ TEST_F(NVFuserTest, FusionViewConcreteDomain_CUDA) {
...
+  FusionExecutor fe;
+  fe.compileFusion(&fusion, {t0});
+
+  // This should fail at the launch time as 14 is not divisible by the
+  // vector word size. The two domains are merged, but they are not
+  // contiguous, so contig indexing is not involved in this case.
+  ASSERT_ANY_THROW(fe.runFusion({t0}));
+}
+
+TEST_F(NVFuserTest, FusionVectorizeInputToOutput_CUDA) {
```
Comment (csarofeen) [inline test/cpp/jit/test_gpu.cpp:21745]: NIT: Can  you stick with the same problem as above? the 14x14 merged and split? I particularly like that case.

Comment (csarofeen) [reply test/cpp/jit/test_gpu.cpp:21745]: never mind posted this before I realized this is testing the misaligned path.

File: torch/csrc/jit/codegen/cuda/lower_validation.cpp [hunk 9]
```diff
@@ -261,6 +261,35 @@ class VectorizeValidator : public OptInDispatch {
     domains_.insert(m->inner());
   }
 
+  // For the producer tensor, it's indexed first by transformed like
+  // the consumer. So, to find its contig merged domain, use the
+  // consumer TensorDomain with the producer contiguity info.
+  static std::vector<bool> getProducerContiguity(
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_validation.cpp]: Nit: Naming `getProducerContiguity` -> `mapProducerContiguity` or `mapContiguityP2C`

File: torch/csrc/jit/codegen/cuda/executor_utils.cpp [hunk 10]
```diff
@@ -400,10 +393,6 @@ c10::optional<std::pair<bool, int>> getVectorizedFusionInputOutput(
   if (producer_tv->isFusionInput()) {
     // Assume consumer_tv is contiguous and only check the producer_tv
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/executor_utils.cpp]: Is this still an assumption here?

File: torch/csrc/jit/codegen/cuda/executor_utils.cpp [hunk 11]
```diff
@@ -536,41 +545,55 @@ void validateAlignedVectorizeExtents(
...
   // domain must have stride 1.
   int64_t cur_contig_stride = 1;
   bool still_rightmost = true;
   for (auto i = aten_tensor.ndimension() - 1; i >= 0; --i) {
-    auto stride = aten_tensor.strides().at(i);
-    auto size = aten_tensor.sizes().at(i);
+    const auto stride = aten_tensor.strides().at(i);
     // If this domain is contiguous, then not necessary to check the
     // stride. Otherwise, stride must be 1 if it's rightmost or
     // divisible by word_size.
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/executor_utils.cpp:576]: Isn't this only true if we merge all contiguous dimensions? Shouldn't this take into consideration the root domains used in the contig_id found from the vectorized id?

File: torch/csrc/jit/codegen/cuda/executor_utils.cpp (hunk 10 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/executor_utils.cpp]: No, thanks for pointing it out.

File: torch/csrc/jit/codegen/cuda/executor_utils.cpp (hunk 11 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/executor_utils.cpp:576]: Hmm, I thought these checks here are sufficient given that we also validate extents with `validateAlignedVectorizeExtents`. Do you have any specific example in your mind?

**Comment by csarofeen** [comment]
Okay, I needed to sort everything going on and I'm including my raw notes below:

- At lowering:
- ```validateAndCollectVectorizeInfo``` is called on the fusion
- This checks all the expressions and grabs any TV with a vectorized dim and calls
- ```VectorizeValidator::validate(tv)```
- which checks the vectorized dimension is a constant value of appropriate size
- creates a `VectorizeValidator` instance with the vectorized id and runs the exprs to produce that vectorized id
- this propagates the vectorized id to its inner most root which is validated to be the inner most dimension of the consumer root domain (Shouldn't this stop at rfactor? Should this also check the producer?)
- it also collects all the ID's involved in expressions to propagate this vectorized id which is only used for misaligned vectorize
- Information is saved off in `VectorizeValidator::validate(tv)` which also calls `VectorizeValidator::fillVectorizedContigRootDomains`
- `VectorizeValidator::fillVectorizedContigRootDomains` fills some information about the contiguous dimension, primarily the least contiguous merge involving the vectorized id from the producer and consumer. Since the vectorized dimension is known to be from the inner most position of the consumer, this contiguous merge must involve the inner most dimension of the consumer.

So for the lowering I think the only concern I have is that the inner most dimension should be validated at rfactor, not at the root domain. We do not validate that the vectorized dimension is the inner most dimension of the producer because we only support vectorized operations on unary ops so no transpose, which means we assume that the inner most dimension of the consumer maps to the inner most dimension of the producer. This probably should be noted.

For the runtime checks:
`validateVectorizedTensors` has 3 calls `validateAlignedVectorizedTensors` `validateMisalignedVectorizedTensors` and `validateVectorizedSplits`

Getting rid of the latter two which isn't important for this PR:
`validateVectorizedSplits` uses `NonDivisibleSplitInfo` has the comment:
//! Predicating non-divisible split domains does not work if split
//! output domains are vectorized where ParallelType::Vectorize is
//! applied to an inner domain of splits. If it's non-divisible,
//! predicating the input domain of the non-divisible split results in
//! a vectoried operation is predicated out entirely since we do not
//! generate a fall-back non-vectorized else path. Runtime check is
//! done for those domains.
From what I understand this is a validation mechanism guaranteeing that we have divisible splits or that we have predicated appropriately.


I'm ignoring `validateMisalignedVectorizedTensors` because we're not focused on misaligned here

So for `validateAlignedVectorizedTensors`:
First thing is done is to cache: `executor_utils::getVectorizedTensorValidationInfo` with the executor's caching system. This is getting first routed to: `getVectorizedFusionInputOutput` which simply marks if  the producer is a fusion input, and notes its position, and marks if the consumer is a fusion output and marks its position. (Returns potentially 2 pairs of <bool, int> bool = true is the producer is a fusion input, bool = false is the consumer is a fusion output.

So `getVectorizedTensorValidationInfo` just returns this information gathered.
 
This information is ignored for now and we go over the lowered information and call `validateAlignedVectorizeExtents`. This runs through the contiguous root set (the root domains part of the contiguous merge found in `VectorizeValidator::fillVectorizedContigRootDomains`. It makes sure the extent of these multiplied is a multiple of the vector size.

Then `validateAlignedVectorizedTensors` calls `validateAlignedVectorizedFusionInputOutput` which goes from the inner most stride to the outer most stride, if the stride is 1 it's ignored, if the stride is "still rightmost" meaning the size is still 1 it's ignored, if the stride remains to be fully contiguous it's ignored.

**Comment by csarofeen** [comment]
There was a few things I had trouble putting together which was why I put the comment: https://github.com/csarofeen/pytorch/pull/1508#discussion_r828015490

There's logic that's a bit all over the place. So the `contigId` logic, makes sure that ID's in a merge are from root ID's that are contiguous and next to each other.  The `VectorizeValidator` makes sure the vector dimension maps to the inner most root domain. It then uses the `contigId` of the vectorized dim to propagate this up to neighbors, neighbors can only inch from right to left otherwise it's not a `contigId` so we know which dimensions we have to validate. `validateAlignedVectorizeExtents` makes sure those inner most contig ID's are a multiple of the vector size. `validateAlignedVectorizedFusionInputOutput` will ignore all contiguous dimensions, and check strides of non-contiguous dimensions. This is helpful because contiguous strides (relative to a fully contiguous tensor) is checked by `validateAlignedVectorizeExtents` but for a non-contiguous dimension someone could throw us something really odd like 7.

File: torch/csrc/jit/codegen/cuda/executor_utils.cpp (hunk 11 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/executor_utils.cpp:576]: Resolved based on my notes below. Leaving open because it can be really hard to follow this logic because it's a bit all over the code base, so leaving this open plus my notes below.

**Review by csarofeen — APPROVED** [review]
Approved. Thanks for doing this, the logic was a lot trickier to get through than I thought it would be, but seems like everything is good now!

**Comment by naoyam** [comment]
Thanks so much for summarizing the current state of vectorization. 

Filed an issue on checking against rfactor domains, #1523. I'll work on it later.
